File: Src/java/poc/translator/src/main/java/org/cqframework/cql/poc/translator/preprocessor/LibraryInfo.java
package org.cqframework.cql.poc.translator.preprocessor;

import org.cqframework.cql.poc.translator.model.*;

import java.util.*;

public class LibraryInfo {
    private String libraryName;
    private String version;

    private final Map<String, UsingDefinitionInfo> usingDefinitions;
    private final Map<String, IncludeDefinitionInfo> includeDefinitions;
    private final Map<String, ValuesetDefinitionInfo> valuesetDefinitions;
    private final Map<String, ParameterDefinitionInfo> parameterDefinitions;
    private final Map<String, LetStatementInfo> letStatements;
    private final Map<String, FunctionDefinitionInfo> functionDefinitions; // TODO: Overloads...
    private final Map<String, RetrieveDefinitionInfo> retrieveDefinitions;

    public LibraryInfo() {
        usingDefinitions = new LinkedHashMap<>();
        includeDefinitions = new LinkedHashMap<>();
        valuesetDefinitions = new LinkedHashMap<>();
        parameterDefinitions = new LinkedHashMap<>();
        letStatements = new LinkedHashMap<>();
        functionDefinitions = new LinkedHashMap<>();
        retrieveDefinitions = new LinkedHashMap<>();
    }

    public String getLibraryName() {
        return libraryName;
    }

    public void setLibraryName(String libraryName) {
        this.libraryName = libraryName;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public LibraryInfo withLibraryName(String value) {
        setLibraryName(value);
        return this;
    }

    public LibraryInfo withVersion(String value) {
        setVersion(value);
        return this;
    }

    public void addUsingDefinition(UsingDefinitionInfo usingDefinition) {
        usingDefinitions.put(usingDefinition.getName(), usingDefinition);
    }

    public UsingDefinitionInfo resolveModelReference(String identifier) {
        return usingDefinitions.get(identifier);
    }

    public void addIncludeDefinition(IncludeDefinitionInfo includeDefinition) {
        includeDefinitions.put(includeDefinition.getLocalName(), includeDefinition);
    }

    public IncludeDefinitionInfo resolveLibraryReference(String identifier) {
        return includeDefinitions.get(identifier);
    }

    public String resolveLibraryName(String identifier) {
        IncludeDefinitionInfo includeDefinition = resolveLibraryReference(identifier);
        if (includeDefinition != null) {
            return includeDefinition.getLocalName();
        }

        return null;
    }

    public void addParameterDefinition(ParameterDefinitionInfo parameterDefinition) {
        parameterDefinitions.put(parameterDefinition.getName(), parameterDefinition);
    }

    public ParameterDefinitionInfo resolveParameterReference(String identifier) {
        return parameterDefinitions.get(identifier);
    }

    public String resolveParameterName(String identifier) {
        ParameterDefinitionInfo parameterDefinition = resolveParameterReference(identifier);
        if (parameterDefinition != null) {
            return parameterDefinition.getName();
        }

        return null;
    }

    public void addValuesetDefinition(ValuesetDefinitionInfo valuesetDefinition) {
        valuesetDefinitions.put(valuesetDefinition.getName(), valuesetDefinition);
    }

    public ValuesetDefinitionInfo resolveValuesetReference(String identifier) {
        return valuesetDefinitions.get(identifier);
    }

    public String resolveValuesetName(String identifier) {
        ValuesetDefinitionInfo valuesetDefinition = resolveValuesetReference(identifier);
        if (valuesetDefinition != null) {
            return valuesetDefinition.getName();
        }

        return null;
    }

    public void addLetStatement(LetStatementInfo letStatement) {
        letStatements.put(letStatement.getName(), letStatement);
    }

    public LetStatementInfo resolveExpressionReference(String identifier) {
        return letStatements.get(identifier);
    }

    public String resolveExpressionName(String identifier) {
        LetStatementInfo letStatement = resolveExpressionReference(identifier);
        if (letStatement != null) {
            return letStatement.getName();
        }

        return null;
    }

    public void addFunctionDefinition(FunctionDefinitionInfo functionDefinition) {
        functionDefinitions.put(functionDefinition.getName(), functionDefinition);
    }

    public FunctionDefinitionInfo resolveFunctionReference(String identifier) {
        return functionDefinitions.get(identifier);
    }

    public String resolveFunctionName(String identifier) {
        FunctionDefinitionInfo functionDefinition = resolveFunctionReference(identifier);
        if (functionDefinition != null) {
            return functionDefinition.getName();
        }

        return null;
    }

    public void addRetrieveDefinition(RetrieveDefinitionInfo retrieveDefinition) {
        retrieveDefinitions.put(retrieveDefinition.getName(), retrieveDefinition);
    }

    public RetrieveDefinitionInfo resolveRetrieveReference(String identifier) {
        return retrieveDefinitions.get(identifier);
    }
}

File: Src/java/cql-to-elm/src/commonMain/kotlin/org/cqframework/cql/cql2elm/preprocessor/LibraryInfo.kt
package org.cqframework.cql.cql2elm.preprocessor

import org.antlr.v4.kotlinruntime.misc.Interval
import org.antlr.v4.kotlinruntime.tree.ParseTree
import org.cqframework.cql.gen.cqlParser.ContextDefinitionContext
import org.cqframework.cql.gen.cqlParser.LibraryDefinitionContext

@Suppress("TooManyFunctions")
class LibraryInfo(
    val namespaceName: String? = null,
    val libraryName: String? = null,
    val version: String? = null,
    override val definition: LibraryDefinitionContext? = null
) : BaseInfo(definition) {

    var defaultUsingDefinition: UsingDefinitionInfo? = null

    private val usingDefinitions = LinkedHashMap<String, UsingDefinitionInfo>()
    private val includeDefinitions = LinkedHashMap<String, IncludeDefinitionInfo>()
    private val codesystemDefinitions = LinkedHashMap<String, CodesystemDefinitionInfo>()
    private val valuesetDefinitions = LinkedHashMap<String, ValuesetDefinitionInfo>()
    private val codeDefinitions = LinkedHashMap<String, CodeDefinitionInfo>()
    private val conceptDefinitions = LinkedHashMap<String, ConceptDefinitionInfo>()
    private val parameterDefinitions = LinkedHashMap<String, ParameterDefinitionInfo>()
    private val expressionDefinitions = LinkedHashMap<String, ExpressionDefinitionInfo>()
    private val functionDefinitions = LinkedHashMap<String, MutableList<FunctionDefinitionInfo>>()
    private val contextDefinitions = ArrayList<ContextDefinitionInfo>()
    private val definitions = HashMap<Interval, BaseInfo>()

    private fun addDefinition(definition: BaseInfo) {
        if (definition.definition != null) {
            val sourceInterval = definition.definition?.sourceInterval
            if (sourceInterval != null) {
                definitions[sourceInterval] = definition
            }
        }
    }

    fun addUsingDefinition(usingDefinition: UsingDefinitionInfo) {
        // First using definition encountered is "preferred", meaning it will resolve as the default
        // model info
        if (defaultUsingDefinition == null) {
            defaultUsingDefinition = usingDefinition
        }
        usingDefinitions[usingDefinition.name] = usingDefinition
        addDefinition(usingDefinition)
    }

    fun resolveModelReference(identifier: String): UsingDefinitionInfo? {
        return usingDefinitions[identifier]
    }

    val defaultModelName: String
        get() {
            val usingDefinitionInfo =
                defaultUsingDefinition
                    ?: throw IllegalArgumentException(
                        "Could not determine a default model because no usings have been defined."
                    )
            return usingDefinitionInfo.name
        }

    fun addIncludeDefinition(includeDefinition: IncludeDefinitionInfo) {
        includeDefinitions[includeDefinition.localName] = includeDefinition
        addDefinition(includeDefinition)
    }

    fun resolveLibraryReference(identifier: String): IncludeDefinitionInfo? {
        return includeDefinitions[identifier]
    }

    fun resolveLibraryName(identifier: String): String? {
        val includeDefinition = resolveLibraryReference(identifier)
        return includeDefinition?.localName
    }

    fun addParameterDefinition(parameterDefinition: ParameterDefinitionInfo) {
        parameterDefinitions[parameterDefinition.name] = parameterDefinition
        addDefinition(parameterDefinition)
    }

    fun resolveParameterReference(identifier: String): ParameterDefinitionInfo? {
        return parameterDefinitions[identifier]
    }

    fun resolveParameterName(identifier: String): String? {
        val parameterDefinition = resolveParameterReference(identifier)
        return parameterDefinition?.name
    }

    fun addCodesystemDefinition(codesystemDefinition: CodesystemDefinitionInfo) {
        codesystemDefinitions[codesystemDefinition.name] = codesystemDefinition
        addDefinition(codesystemDefinition)
    }

    fun resolveCodesystemReference(identifier: String): CodesystemDefinitionInfo? {
        return codesystemDefinitions[identifier]
    }

    fun addValuesetDefinition(valuesetDefinition: ValuesetDefinitionInfo) {
        valuesetDefinitions[valuesetDefinition.name] = valuesetDefinition
        addDefinition(valuesetDefinition)
    }

    fun resolveValuesetReference(identifier: String): ValuesetDefinitionInfo? {
        return valuesetDefinitions[identifier]
    }

    fun resolveValuesetName(identifier: String): String? {
        val valuesetDefinition = resolveValuesetReference(identifier)
        return valuesetDefinition?.name
    }

    fun addCodeDefinition(codeDefinition: CodeDefinitionInfo) {
        codeDefinitions[codeDefinition.name] = codeDefinition
        addDefinition(codeDefinition)
    }

    fun resolveCodeReference(identifier: String): CodeDefinitionInfo? {
        return codeDefinitions[identifier]
    }

    fun addConceptDefinition(conceptDefinition: ConceptDefinitionInfo) {
        conceptDefinitions[conceptDefinition.name] = conceptDefinition
        addDefinition(conceptDefinition)
    }

    fun resolveConceptReference(identifier: String): ConceptDefinitionInfo? {
        return conceptDefinitions[identifier]
    }

    fun addExpressionDefinition(letStatement: ExpressionDefinitionInfo) {
        expressionDefinitions[letStatement.name] = letStatement
        addDefinition(letStatement)
    }

    fun resolveExpressionReference(identifier: String): ExpressionDefinitionInfo? {
        return expressionDefinitions[identifier]
    }

    fun resolveExpressionName(identifier: String): String? {
        val letStatement = resolveExpressionReference(identifier)
        return letStatement?.name
    }

    fun addFunctionDefinition(functionDefinition: FunctionDefinitionInfo) {
        var infos = functionDefinitions[functionDefinition.name]
        if (infos == null) {
            infos = ArrayList()
            functionDefinitions[functionDefinition.name] = infos
        }
        infos.add(functionDefinition)
        addDefinition(functionDefinition)
    }

    fun resolveFunctionReference(identifier: String): List<FunctionDefinitionInfo>? {
        return functionDefinitions[identifier]
    }

    fun resolveFunctionName(identifier: String): String? {
        val functionDefinitions = resolveFunctionReference(identifier)
        if (functionDefinitions != null) {
            for (functionInfo in functionDefinitions) {
                return functionInfo.name
            }
        }
        return null
    }

    fun addContextDefinition(contextDefinition: ContextDefinitionInfo) {
        contextDefinitions.add(contextDefinition)
        addDefinition(contextDefinition)
    }

    fun resolveContext(ctx: ContextDefinitionContext): ContextDefinitionInfo? {
        for (cd in contextDefinitions) {
            if (ctx.sourceInterval == cd.definition.sourceInterval) {
                return cd
            }
        }
        return null
    }

    fun resolveDefinition(pt: ParseTree): BaseInfo? {
        return definitions[pt.sourceInterval]
    }
}
